lib/ansible/modules/cloud/misc/qemu_img.py

```diff
@@ -0,0 +1,221 @@
+- qemu_img:
+    dest: /tmp/testimg
+    size: 5M
+    format: raw
+
+# Enlarge the image to 6G.
+- qemu_img:
+    dest: /tmp/testimg
+    size: 6G
+    format: qcow2
+
+# Shrink the image by 3G
+    code: qemu_img
+    dest: /tmp/testing
+    size: -3G
+    shrink: yes
+    format: qcow2
+
+# Remove the image
+- qemu_img:
+    dest: /tmp/testimg
+    state: absent
+'''
+
+RETURN = '''
+# create qemu image
+present:
+  type: string
+  sample: "success"
+  returned: success
+# remove qemu image
+absent:
+  type: string
+  sample: "success"
+  returned: success
+# resize qemu image
+resize:
+  type: string
+  sample: "success"
+  returned: success
+'''
+
+import os
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            dest = dict(type='str', required=True),
+            options = dict(type='str', default='preallocation=metadata'),
+            format = dict(type='str', default='qcow2'),
+            size = dict(type='str'),
+            grow= dict(type="bool", default=True),
+            shrink = dict(type="bool", default=False),
+            state = dict(type='str', choices=['absent', 'present'], default='present'),
+        ),
+        supports_check_mode = True,
+    )
+
+    result = dict(
+        changed = False,
+    )
+
+    state = module.params['state']
+    dest = module.params['dest']
+    img_format = module.params['format']
+    img_options = module.params['options']
+    size = module.params['size']
+    grow = module.boolean(module.params['grow'])
+    shrink = module.boolean(module.params['shrink'])
+
+    qemu_img = module.get_bin_path('qemu-img', True)
+
+    if state == 'present':
+        if not size:
+            module.fail_json(msg="No size defined, creating a disk image requires a size")
+        if not os.path.exists(dest):
+            if not module.check_mode:
+                if not img_options:
+                    module.run_command('%s create -f %s "%s" %s' % (qemu_img, img_format, dest, size), check_rc=True)
+                else:
+                    module.run_command('%s create -f %s -o "%s" "%s" %s' % (qemu_img, img_format, img_options, dest, size), check_rc=True)
+            result['changed'] = True
+        else:
+            size_unit = size[-1:]
+            if size_unit not in ['T','G','M','k','b','']:
+                module.fail_json(msg="No valid size unit specified")
+            else:
+                if(size_unit == 'T'):
+                    unitmultiplier = 1024*1024*1024*1024
+                elif(size_unit == 'G'):
+                    unitmultiplier = 1024*1024*1024
+                elif(size_unit == 'M'):
+                    unitmultiplier = 1024*1024
+                elif(size_unit == 'k'):
+                    unitmultiplier = 1024
+                else:
+                    unitmultiplier = 1
+            size = size[:-1]
```

Comment:
I have rewritten the code based upon your compact version, however I disagree with making the unit mandatory, given that qemu-img does not require a unit and it states that in very clearly in the documentation, given that the ansible module uses the qemu-img command, I think we need to follow the same logic as qemu-img